rskj-core/src/main/java/co/rsk/core/bc/MiningMainchainViewImpl.java

```diff
@@ -20,101 +20,180 @@
+                addBestAndRebuildFromParent(bestHeader, parentIndex.getAsInt());
+            } else {
+                buildMainchainFromList(Arrays.asList(bestHeader));
+            }
 
-            fillInternalsBlockStore(bestBlock);
-            deleteEntriesOutOfBoundaries(bestBlock.getNumber());
+            deleteEntriesOutOfBoundaries(bestHeader.getNumber());
         }
     }
 
     @Override
-    public List<Block> get() {
+    public List<BlockHeader> get() {
         synchronized (internalBlockStoreReadWriteLock) {
             return Collections.unmodifiableList(mainchain);
         }
     }
 
-    @Override
-    public Block getBestBlock() {
-        synchronized (internalBlockStoreReadWriteLock) {
-            return mainchain.get(0);
-        }
+    /**
+     * Given a new best header and the index of its parent rebuild the mainchain using the index as the pivot point
+     * by discarding all headers subsequent to the it, setting the new header as the tip and refilling with as many
+     * are needed to complete the desired height
+     *
+     * @param bestHeader The best header to be on top of the chain
+     * @param parentIndex List index of the best header's parent
+     */
+    private void addBestAndRebuildFromParent(BlockHeader bestHeader, int parentIndex) {
+        int toSkip = parentIndex + 1;
+
+        List<BlockHeader> commonAncestorChain = mainchain.stream()
+                .skip(toSkip)
+                .collect(Collectors.toList());
+
+        List<BlockHeader> newMainchain = Stream.concat(
+                Arrays.asList(bestHeader).stream(),
+                commonAncestorChain.stream())
+                .collect(Collectors.toList());
+
+        buildMainchainFromList(newMainchain);
     }
 
-    @Override
-    public Block getBlockByNumber(long number) {
-        return blockchain.getBlockByNumber(number);
-    }
+    /**
+     * Given a source list, take it as the new mainchain and refill it with as many block headers are needed to reach
+     * the desired depth/height
+     *
+     * @param sourceList
+     */
+    private void buildMainchainFromList(List<BlockHeader> sourceList) {
+        int sourceSize = sourceList.size();
+
+        if (sourceSize == 0 || height < sourceSize) {
+            return;
+        }
+
+        BlockHeader lastHeader = sourceList.get(sourceSize - 1);
+
+        List<BlockHeader> missingHeaders = retrieveAncestorsForHeader(lastHeader, height - sourceSize);
 
-    private void fillInternalsBlockStore(Block bestBlock) {
-        List<Block> newBlockchain = new ArrayList<>(height);
-        Block currentBlock = bestBlock;
-        for(int i = 0; i < height; i++) {
-            newBlockchain.add(currentBlock);
+        for (BlockHeader header : missingHeaders) {
+            if(!blocksByHash.containsKey(header.getHash())) {
+                addHeaderToMaps(header);
+            }
+        }
+
+        mainchain = Stream.concat(sourceList.stream(), missingHeaders.stream())
```

Comment:
expensive copying of items when missingHeaders is just a few items. 
In the future a circular buffer would work reduce most unnecessary element copying. Also you can reduce copying by reversing the list (best header last) and moving the "first" pointer, with periodic removals of initial items (e.g. every 1000 elements added).

Reply:
It also can be used to remove old headers only once in a while, with reduced cost (amortized).